frontend/catalyst/python_interface/pass_api/passes.py

```diff
@@ -0,0 +1,199 @@
+        def match_and_rewrite(self, op: Operation, rewriter: PatternRewriter) -> None:
+            rewrite_rule(self._pass, op, rewriter)
+
+    return _RewritePattern
+
+
+class PLModulePass(ModulePass):
+    """An xdsl ``ModulePass`` subclass for defining passes."""
+
+    name: ClassVar[str]
+    _rewrite_patterns: ClassVar[dict[Operation, RewritePattern]] = {}
+
+    def __init__(self, recursive: bool = True, greedy: bool = False):
+        self.recursive = recursive
+        self.greedy = greedy
+
+    @property
+    def recursive(self):
+        """Whether or not the rewrite rules should be applied recursively. If ``True``,
+        the rewrite rules will be applied repeatedly until a steady-state is reached.
+        ``True`` by default.
+        """
+        return True
+
+    @property
+    def greedy(self):
+        """Whether or not the rewrite rules should be applied greedily. If ``True``,
+        each iteration of the rewrite rules' application (if ``recursive == True``)
+        will only apply the first rewrite rule that modifies the input module.
+        ``True`` by default."""
+        return True
+
+    @classmethod
+    def rewrite_rule(
+        cls, rule: Callable[["PLModulePass", Operation, PatternRewriter], None]
+    ) -> None:
+        """Register a rewrite rule.
+
+        The rewrite rule must type hint which operation is being rewritten. It must have
+        the following signature:
+
+        .. code-block:: python
+
+            @PLModulePass.rewrite_rule
+            def rewrite_myop(self, op: MyOperation, rewriter: PatternRewriter) -> None:
+                ...
+
+        In the above example, the type hint for the second argument, which is ``op``,
+        is used to determine which xDSL operation is being matched. If not provided,
+        _all_ xDSL operations will be matched.
+
+        Additionally, the type hint can also contain a union of multiple operation types:
+
+        .. code-block:: python
+
+            @PLModulePass.rewrite_rule
+            def rewrite_myop(
+                self, op: MyOperation1 | MyOperation2 | MyOperation3, rewriter: PatternRewriter
+            ) -> None:
+                ...
+
+        .. note::
+
+            If a rewrite rule for the provided operation already exists, the old rule
+            will get overwritten. For rewrite rules that were annotated with a union of
+            multiple operation types, registered rewrite rules for all of the operations
+            in the union will be overwritten.
+
+        Args:
+            hint (type[Operation]): Operation class for which to register the rule
+
+        Returns:
+            Callable: a decorator to register the rewrite rule with the ModulePass
+        """
+        # xdsl.pattern_rewriter.op_type_rewrite_pattern was used as a reference to
+        # implement the type hint collection. Source:
+        # https://github.com/xdslproject/xdsl/blob/main/xdsl/pattern_rewriter.py
```

Comment:
xDSL uses the [Apache 2.0 license](https://github.com/xdslproject/xdsl/blob/main/LICENSE), so this should be fine, right? The functionality mentioned in this comment is not something that does anything actually related to xDSL or compilation either.